lib/protobuf/json/encode.ex

```diff
@@ -0,0 +1,129 @@
+defmodule Protobuf.JSON.Encode do
+  @moduledoc """
+  Internal JSON encoding API.
+  """
+
+  alias Protobuf.MessageProps
+
+  @type encode_opt ::
+          {:use_proto_names, boolean}
+          | {:use_enum_numbers, boolean}
+          | {:emit_unpopulated, boolean}
+
+  @compile {:inline, encode_field: 3, encode_key: 2, maybe_repeat: 3}
+
+  @spec encode(struct, MessageProps.t(), [encode_opt]) :: map
+  def encode(%_{} = struct, %MessageProps{} = message_props, opts \\ []) do
+    regular = encode_regular_fields(struct, message_props, opts)
+    oneofs = encode_oneof_fields(struct, message_props, opts)
+
+    :maps.from_list(regular ++ oneofs)
+  end
+
+  defp encode_regular_fields(struct, %{field_props: field_props}, opts) do
+    for {_field_num, %{name_atom: name, oneof: nil} = prop} <- field_props,
+        %{^name => value} = struct,
+        opts[:emit_unpopulated] || !default?(prop, value) do
+      encode_field(prop, value, opts)
+    end
+  end
+
+  # TODO: is %{field: {:tag, default_value}} valid? should it be treated
+  # as %{field: nil} and skipped too?
+  defp encode_oneof_fields(struct, message_props, opts) do
+    %{field_tags: field_tags, field_props: field_props, oneof: oneofs} = message_props
+
+    for {oneof_name, _index} <- oneofs,
+        tag_and_value = Map.get(struct, oneof_name) do
+      {tag, value} = tag_and_value
+      prop = field_props[field_tags[tag]]
+      encode_field(prop, value, opts)
+    end
+  end
+
+  # TODO: handle invalid values? check types?
```

Comment:
When structs are built straight out of a binary using `Protobuf.decode` we know for sure they are well-formed and will encode to JSON correctly. Manipulated structs however may crash this at some point. I wonder how strict we should be in this regard without compromising performance.

Reply:
I think we can delve deeper into this on a separate Issue/PR. As far as my research goes, all the dynamic languages implementations do type checking at the moment the fields are set, which seems like a good idea because if something is wrong with our data we get early feedback by failing fast, as opposed to doing it at encoding/serialization time. In the end I think the challenge will be finding a sensible way of doing that without compromising  too much the niceties of working with structs naturally.